server/src/main/java/org/elasticsearch/search/retriever/CompoundRetrieverBuilder.java

```diff
@@ -0,0 +1,225 @@
+
+    protected final int rankWindowSize;
+    protected final List<RetrieverSource> innerRetrievers;
+
+    protected CompoundRetrieverBuilder(List<RetrieverSource> innerRetrievers, int rankWindowSize) {
+        this.rankWindowSize = rankWindowSize;
+        this.innerRetrievers = innerRetrievers;
+    }
+
+    @SuppressWarnings("unchecked")
+    public T addChild(RetrieverBuilder retrieverBuilder) {
+        innerRetrievers.add(new RetrieverSource(retrieverBuilder, null));
+        return (T) this;
+    }
+
+    /**
+     * Returns a clone of the original retriever, replacing the sub-retrievers with
+     * the provided {@code newChildRetrievers}.
+     */
+    public abstract T clone(List<RetrieverSource> newChildRetrievers);
+
+    /**
+     * Combines the provided {@code rankResults} to return the final top documents.
+     */
+    public abstract RankDoc[] combineInnerRetrieverResults(List<ScoreDoc[]> rankResults);
+
+    @Override
+    public final boolean isCompound() {
+        return true;
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public final RetrieverBuilder rewrite(QueryRewriteContext ctx) throws IOException {
+        if (ctx.getPointInTimeBuilder() == null) {
+            throw new IllegalStateException("PIT is required");
+        }
+
+        // Rewrite prefilters
+        boolean hasChanged = false;
+        var newPreFilters = rewritePreFilters(ctx);
+        hasChanged |= newPreFilters != preFilterQueryBuilders;
+
+        // Rewrite retriever sources
+        List<RetrieverSource> newRetrievers = new ArrayList<>();
+        for (var entry : innerRetrievers) {
+            RetrieverBuilder newRetriever = entry.retriever.rewrite(ctx);
+            if (newRetriever != entry.retriever) {
+                newRetrievers.add(new RetrieverSource(newRetriever, null));
+                hasChanged |= true;
+            } else {
+                var sourceBuilder = entry.source != null
+                    ? entry.source
+                    : createSearchSourceBuilder(ctx.getPointInTimeBuilder(), newRetriever);
+                var rewrittenSource = sourceBuilder.rewrite(ctx);
+                newRetrievers.add(new RetrieverSource(newRetriever, rewrittenSource));
+                hasChanged |= rewrittenSource != entry.source;
+            }
+        }
+        if (hasChanged) {
+            return clone(newRetrievers);
+        }
+
+        // execute searches
+        final SetOnce<RankDoc[]> results = new SetOnce<>();
+        final MultiSearchRequest multiSearchRequest = new MultiSearchRequest();
+        for (var entry : innerRetrievers) {
+            SearchRequest searchRequest = new SearchRequest().source(entry.source);
+            // The can match phase can reorder shards, so we disable it to ensure the stable ordering
+            // TODO: does this break if we query frozen tier? should we update/remove this
+            searchRequest.setPreFilterShardSize(Integer.MAX_VALUE);
```

Comment:
@jimczi could you please take a look at this and share your thoughts? Would this have any undesired effects on the can match phase (e.g. for frozen tiers) ?

Reply:
>  For example, if a request is sorted by timestamp in descending order, the shards with the latest timestamp will be processed first. However, this optimization doesn't help in practice because we still need to wait for all shards to finish.

Thanks for the clarification @jimczi ! An additional question to help me better understand: Is the main issue that may arise here that if for some reason some shards are re-ordered for some sub-retrievers, then the `shardIndex` part of `RankDoc` would not be consistent and comparable among different result sets?

Reply:
Exactly, so we need to ensure that rank docs always refer to same `shardIndex`.`shardIndex` depends on the ordering of the shards in the request.

Reply:
Added a quick take on this in 06ec5387029b0f101c8752915420962210cf82b7. If we're ok with something like this as a starting point, I'm happy to move it to a separate PR.

Reply:
Sounds good @pmpailis!

Reply:
Created https://github.com/elastic/elasticsearch/pull/113118/

Reply:
[Closing the above](https://github.com/elastic/elasticsearch/pull/113118#issuecomment-2361185917) PR in favor of disabling can-match completely.  The main issue was that even if we disable sorting shards, skipping some shards for range queries could still affect `shardIndex` computation and how we merge `RankDoc` results

Reply:
Skipping can_match will hurt frozen & large search requests significantly. Just so we are aware. 

Is there no way to resolve this without removing this significant optimization?

Reply:
Yeap, that's really a nasty side-effect. One option is look into changing the encoding of the `RankDoc.RankKey` to capture additional info aside from just `shardIndex` to move away from this limitation and make use of all optimizations, but that would probably require a significant refactoring. So, maybe this could be in a follow up PR to avoid adding even more functionality and logic to this one (it's already pretty huge as it is).

Reply:
> Skipping can_match will hurt frozen & large search requests significantly

We still execute the can match phase for every shard before entering the query phase on the frozen node. So I am not sure that it will hurt search requests so significantly (See https://github.com/elastic/elasticsearch/blob/7fc0a41083c044ddc4d564f7b77a5f4e591e460e/server/src/main/java/org/elasticsearch/search/SearchService.java#L547).
It's also not the main use case for this feature but I agree that we have room for improvements here. We could for instance create the PIT using the global filters of the request to ensure that we don't select too much shards when creating it. All this can be done in a follow up imo, let's start simple.